sdks/java/extensions/timeseries/src/main/java/org/apache/beam/sdk/extensions/timeseries/FillGaps.java

```diff
@@ -0,0 +1,532 @@
+
+    // A timer map used to fill potential gaps. Each logical "window" will have a separate timer
+    // which will be cleared if an element arrives in that window. This way the timer will only fire
+    // if there is a gap, at which point it will fill the gap.
+    @TimerFamily("gapTimers")
+    @SuppressWarnings({"UnusedVariable"})
+    private final TimerSpec gapFillingTimersSpec = TimerSpecs.timerMap(TimeDomain.EVENT_TIME);
+
+    // Timers used to garbage collect state.
+    @TimerFamily("gcTimers")
+    @SuppressWarnings({"UnusedVariable"})
+    private final TimerSpec gcTimersSpec = TimerSpecs.timerMap(TimeDomain.EVENT_TIME);
+
+    // Keep track of windows already seen. In the future we can replace this with OrderedListState.
+    // Keyed by window end timestamp (which is 1ms greater than the window max timestamp).
+    @StateId("seenBuckets")
+    @SuppressWarnings({"UnusedVariable"})
+    private final StateSpec<ValueState<SortedMap<Instant, TimestampedValue<ValueT>>>>
+        seenBucketsSpec;
+
+    // For every window, keep track of how long the filled gap is in buckets. If a window was
+    // populated by a received element - i.e.
+    // it's not
+    // a gap fill - then there is no value in this map for that window.
+    // Keyed by window end timestamp (which is 1ms greater than the window max timestamp).
+    @StateId("gapDurationMap")
+    @SuppressWarnings({"UnusedVariable"})
+    private final StateSpec<ValueState<SortedMap<Instant, Long>>> gapDurationSpec;
+
+    FillGapsDoFn(
+        FixedWindows bucketWindows,
+        Coder<ValueT> valueCoder,
+        Instant stopTime,
+        long maxGapFillBuckets,
+        SerializableBiFunction<
+                TimestampedValue<ValueT>, TimestampedValue<ValueT>, TimestampedValue<ValueT>>
+            mergeValues,
+        @Nullable SerializableFunction<PropagateData<ValueT>, ValueT> propagateFunction) {
+      this.bucketWindows = bucketWindows;
+      this.gcWindows = FixedWindows.of(bucketWindows.getSize().multipliedBy(GC_EVERY_N_BUCKETS));
+      this.stopTime = stopTime;
+      this.maxGapFillBuckets = maxGapFillBuckets;
+      this.seenBucketsSpec =
+          StateSpecs.value(
+              SortedMapCoder.of(InstantCoder.of(), TimestampedValueCoder.of(valueCoder)));
+      this.gapDurationSpec =
+          StateSpecs.value(SortedMapCoder.of(InstantCoder.of(), VarLongCoder.of()));
+      this.mergeValues = mergeValues;
+      this.propagateFunction = propagateFunction;
+    }
+
+    @ProcessElement
+    public void process(
+        @Element KV<Row, ValueT> element,
+        @Timestamp Instant ts,
+        @TimerFamily("gapTimers") TimerMap gapTimers,
+        @TimerFamily("gcTimers") TimerMap gcTimers,
+        @AlwaysFetched @StateId("seenBuckets")
+            ValueState<SortedMap<Instant, TimestampedValue<ValueT>>> seenBuckets,
+        OutputReceiver<ValueT> output) {
+      if (ts.isAfter(stopTime)) {
```

Comment:
What does "absolute" add? stopTime seems clear enough

Reply:
TTL is also a stop indicator, but I believe in this case this value refers to the absolute end time given by the user. This would make it easier to follow the logic.

Reply:
By TTL do you mean the max gap fill? That's not a stop indicator for the pipeline - if more elements show up, we'll start filling again.

I just don't see that absolute adds anything here. Stop time is stop time.